components/camera/videosource/query.go

```diff
@@ -3,25 +3,137 @@ package videosource
 import (
 	"fmt"
 	"math"
+	"path/filepath"
+	"runtime"
 	"strings"
 	"time"
 
 	"github.com/pion/mediadevices"
 	"github.com/pion/mediadevices/pkg/driver"
 	"github.com/pion/mediadevices/pkg/driver/availability"
-	"github.com/pion/mediadevices/pkg/driver/camera"
+	mediadevicescamera "github.com/pion/mediadevices/pkg/driver/camera"
+	"github.com/pion/mediadevices/pkg/frame"
 	"github.com/pion/mediadevices/pkg/io/video"
 	"github.com/pion/mediadevices/pkg/prop"
 	"github.com/pkg/errors"
 
 	"go.viam.com/rdk/logging"
 )
 
+// minResolutionDimension is set to 2 to ensure proper fitness distance calculation for resolution selection.
+// Setting this to 0 would cause mediadevices' IntRanged.Compare() method to treat all values smaller than ideal
+// as equally acceptable. See https://github.com/pion/mediadevices/blob/c10fb000dbbb28597e068468f3175dc68a281bfd/pkg/prop/int.go#L104
+// Setting it to 1 could theoretically allow 1x1 resolutions. 2 is small enough and even,
+// allowing all real camera resolutions while ensuring proper distance calculations.
+const minResolutionDimension = 2
+
 // Below is adapted from github.com/pion/mediadevices.
 // It is further adapted from gostream's query.go
 // However, this is the minimum code needed for webcam to work, placed in this directory.
 // This vastly improves the debugging and feature development experience, by not over-DRY-ing.
 
+// makeConstraints is a helper that returns constraints to mediadevices in order to find and make a video source.
+// Constraints are specifications for the video stream such as frame format, resolution etc.
+func makeConstraints(conf *WebcamConfig, logger logging.Logger) mediadevices.MediaStreamConstraints {
+	return mediadevices.MediaStreamConstraints{
+		Video: func(constraint *mediadevices.MediaTrackConstraints) {
+			if conf.Width > 0 {
+				constraint.Width = prop.IntExact(conf.Width)
+			} else {
+				constraint.Width = prop.IntRanged{Min: minResolutionDimension, Ideal: 640, Max: 4096}
+			}
+
+			if conf.Height > 0 {
+				constraint.Height = prop.IntExact(conf.Height)
+			} else {
+				constraint.Height = prop.IntRanged{Min: minResolutionDimension, Ideal: 480, Max: 2160}
+			}
+
+			if conf.FrameRate > 0.0 {
+				constraint.FrameRate = prop.FloatExact(conf.FrameRate)
+			} else {
+				constraint.FrameRate = prop.FloatRanged{Min: 0.0, Ideal: 30.0, Max: 140.0}
+			}
+
+			if conf.Format == "" {
```

Comment:
[possible nit] I know this was just moved over, but, is the pixel format support not exposed by mediadevices somehow? would be nice not to have to maintain this against mediadevices changes